spec.md

```diff
@@ -811,7 +810,11 @@ A function value represents a function defined in Starlark.
 Its [type](#type) is `"function"`.
 A function value used in a Boolean context is always considered true.
 
-Function definitions may not be nested.
+Functions defined by a [`def` statement](#function-definitions) are named;
+functions defined by a [`lambda` expression](#lambda-expressions) are anonymous.
+
+Function definitions may be nested, and an inner function may refer 
+to a local variable of an outer function.
```

Comment:
Probably worth clarifying that:

* it is not possible to assign to a captured variable (unlike Python with `nonlocal` keyword)
* captured values can be mutated until captured values are frozen (similar to functions capturing globals)

I have mixed feelings about this proposal.

I have never seen any good code with nested def statements.

Lambdas are somewhat more useful, but anonymous functions in 99% of cases are used in parameters to map and filter functions, which is not the issue with Python/Starlark list comprehensions expressions.

But I heard some people want both nested defs and lambdas.

Reply:
> Probably worth clarifying that ...

Done, x2


> I have mixed feelings about this proposal.

My two concerns are that some people will try to write Scheme or Haskell in Starlark, and that the memory footprint of a struct containing only closures (to simulate an opaque object) is greater than necessary. (It's approximately 15 words per closure in the Java implementation.) The first problem may be ameliorated by the lack of recursion, and the second can be addressed by optimizations (e.g. combining defaults and freevars Tuples into a single array; grouping all the per-module stuff behind a single reference).


> I have never seen any good code with nested def statements.

I don't know what code you've seen, but I've seen tons of good and bad code with nested def statements, and in the absence of classes (the other way to combine code+data), they provide a much needed form of expressiveness. They also help to avoid polluting the toplevel namespace with helper functions that could be local.